pkg/component/controller/metrics.go

```diff
@@ -0,0 +1,359 @@
+
+type job struct {
+	scrapeURL     string
+	name          string
+	hostname      string
+	clusterConfig *v1beta1.ClusterConfig
+	log           *logrus.Entry
+	scrapeClient  *http.Client
+	restClient    rest.Interface
+}
+
+func (m *Metrics) newJob(name, scrapeURL string) (*job, error) {
+	certFile := path.Join(m.K0sVars.CertRootDir, "admin.crt")
+	keyFile := path.Join(m.K0sVars.CertRootDir, "admin.key")
+
+	httpClient, err := getClient(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	return &job{
+		scrapeURL:    scrapeURL,
+		name:         name,
+		hostname:     m.hostname,
+		log:          m.log,
+		scrapeClient: httpClient,
+		restClient:   m.restClient,
+	}, nil
+}
+
+func (j *job) Run(ctx context.Context) {
+	j.log.Debugf("Running %s job", j.name)
+
+	t := time.NewTicker(time.Second * 30)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-t.C:
+			if j.clusterConfig == nil {
+				continue
+			}
+
+			err := j.collectAndPush(ctx)
+			if err != nil {
+				j.log.Error(err)
+			}
+		}
+	}
+}
+func (j *job) pushURL() string {
+	pushAddress := fmt.Sprintf("/api/v1/namespaces/%s/services/http:%s:http/proxy", namespace, pushGatewayName)
+	return fmt.Sprintf("%s/metrics/job/%s/instance/%s", pushAddress, j.name, j.hostname)
+}
+
+func (j *job) collectAndPush(ctx context.Context) error {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, j.scrapeURL, nil)
+	if err != nil {
+		return fmt.Errorf("error creating GET request for %s: %w", j.scrapeURL, err)
+	}
+
+	resp, err := j.scrapeClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("error collecting metrics from %s: %w", j.scrapeURL, err)
+	}
+	defer resp.Body.Close()
+
+	res := j.restClient.Post().AbsPath(j.pushURL()).Body(resp.Body).Do(ctx)
+	if res.Error() != nil {
+		return fmt.Errorf("error sending POST request for job %s: %w", j.name, res.Error())
+	}
+	return nil
+}
+
+func getClient(certFile, keyFile string) (*http.Client, error) {
+	transport := http.DefaultTransport.(*http.Transport).Clone()
+	transport.ResponseHeaderTimeout = time.Minute
+	tlsConfig := &tls.Config{InsecureSkipVerify: true}
+	transport.TLSClientConfig = tlsConfig
+
+	if certFile != "" && keyFile != "" {
+		cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+		if err != nil {
+			return nil, err
+		}
+		tlsConfig.Certificates = []tls.Certificate{cert}
+	}
+	return &http.Client{
+		Transport: transport,
+		Timeout:   time.Minute,
+	}, nil
+}
+
+const pushGatewayTemplate = `
+---
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: {{ .Namespace }}
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  labels:
+    component: "pushgateway"
+    app: k0s-observability
+  name: k0s-pushgateway
+  namespace: {{ .Namespace }}
+  annotations:
+    {}
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  labels:
+    component: "pushgateway"
+    app: k0s-observability
+  name: {{ .Name }}
+rules:
+  []
```

Comment:
Why empty set of rules?

Reply:
Yeah, I just copied a manifest generated by helm. 
I think pushgateway SA should use our default PodSecurityPolicy. 
Maybe we can remove this separate ClusterRole and just assign the `k0s:podsecuritypolicy:default` ClusterRole to the SA. WDYT?

Reply:
Why do we need SA for it in the first place?

Reply:
Yeah, we can just remove a separate SA usage.

Reply:
Removed it